File: graphs/keys_and_rooms.py
# Q:841: https://leetcode.com/problems/keys-and-rooms/

from typing import List

class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = [False] * len(rooms)

        def dfs(room):
            visited[room] = True
            keys = rooms[room]

            for key in keys:
                if not visited[key]:
                    dfs(key)
        
        dfs(0)

        return all(visited)




c = Solution()

rooms1 = [[1],[2],[3],[]]
print(c.canVisitAllRooms(rooms1))

rooms2 = [[1,3],[3,0,1],[2],[0]]
print(c.canVisitAllRooms(rooms2))

File: graphs/keys_and_rooms.js
// Q:841: https://leetcode.com/problems/keys-and-rooms/

/**
 * @param {number[][]} rooms
 * @return {boolean}
 */
const canVisitAllRooms = function(rooms) {  
    // initialize the queue with keys in room zero
    const queue = [0];
    const visitedRooms = new Set([0]);
  
    while(queue.length){
        const current = queue.shift();
        
  
        const keysInRoom = rooms[current];
  
        for (let key of keysInRoom){
            if (!visitedRooms.has(key)){
                queue.push(key);
                visitedRooms.add(key);
            }
        }

    }
  
    return visitedRooms.size === rooms.length;
 };